babel/numbers.py

```diff
@@ -475,6 +491,52 @@ def format_currency(
         decimal_quantization=decimal_quantization)
 
 
+def _format_currency_long_name(
+        number, currency, format=None, locale=LC_NUMERIC, currency_digits=True,
+        format_type='standard', decimal_quantization=True):
+    # Algorithm described here:
+    # https://www.unicode.org/reports/tr35/tr35-numbers.html#Currencies
+    locale = Locale.parse(locale)
+    # Step 1.
+    # There are no examples of items with explicit count (0 or 1) in current
+    # locale data. So there is no point implementing that.
+    # Step 2.
+    if isinstance(number, string_types):
+        plural_category = locale.plural_form(float(number))
+    else:
+        plural_category = locale.plural_form(number)
+
+    # Step 3.
+    try:
+        unit_pattern = locale._data['currency_unit_patterns'][plural_category]
```

Comment:
I mildly like pulling things out of `locale._data` from outside the `locale` object, but we haven't been that strict about it before either.

Reply:
I had a partial go at fixing this in 4c7e9b645cfec72672f836e04d8587ca66ba98c3 - it is at least more consistent now, and the resulting code is much flatter as well.